make/compiler_flags

```diff
@@ -0,0 +1,191 @@
+# LDFLAGS_OPENCL
+# LDLIBS_OPENCL
+
+## MPI specific flags
+# CXXFLAGS_MPI
+# LDFLAGS_MPI
+# LDLIBS_MPI
+
+################################################################################
+# Update compiler flags with operating system specific modifications
+##
+ifeq ($(OS),Windows_NT)
+  CXXFLAGS_WARNINGS ?= -Wall -Wno-unused-function -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-variable -Wno-sign-compare -Wno-unused-local-typedefs
+  CPPFLAGS_GTEST ?= -DGTEST_HAS_PTHREAD=0
+  CPPFLAGS_OS ?= -D_USE_MATH_DEFINES
+  ## Detecting Process Bitness:
+  ## http://blogs.msdn.com/b/david.wang/archive/2006/03/26/howto-detect-process-bitness.aspx
+  ifeq ($(PROCESSOR_ARCHITECTURE)$(PROCESSOR_ARCHITEW6432),x86)
+    CXXFLAGS_OS ?= -m32
+  else
+    CXXFLAGS_OS ?= -m64
+  endif
+
+  ifneq (gcc,$(CXX_TYPE))
+    LDLIBS_OS ?= -static-libgcc
+  else
+    LDLIBS_OS ?= -static-libgcc -static-libstdc++
+  endif
+
+  EXE := .exe
+endif
+
+ifeq ($(OS),Darwin)
+  ifeq (clang,$(CXX_TYPE))
+    CXXFLAGS_OS ?= -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare
+  endif
+endif
+
+ifeq ($(OS),Linux)
+  CPPFLAGS_GTEST ?= -DGTEST_HAS_PTHREAD=0
+  CXXFLAGS_WARNINGS ?= -Wno-sign-compare
+  ifeq (gcc,$(CXX_TYPE))
+    CXXFLAGS_OS ?= -pthread
+  endif
+  ifeq (mingw32-gcc,$(CXX_TYPE))
+    LDLIBS_OS ?= -static -lwinpthread
+    EXE := .exe
+    WINE ?= wine
+  endif
+  ifeq (clang,$(CXX_TYPE))
+    LDLIBS_OS ?= -lpthread
+    LDFLAGS_OS ?= -stdlib=libc++
+  endif
+endif
+
+
+################################################################################
+# Setup GPU
+#
+# Adds the following to CXXFLAGS_OPENCL
+# OPENCL_DEVICE_ID - The ID of the GPU (default: 0)
+# OPENCL_PLATFORM_ID The ID of the OpenCL platform (default: 0)
+# Both IDs can be found through installing and calling clinfo
+
+ifdef STAN_OPENCL
+  ifeq ($(OS),Darwin)  # Mac OS X
+    LDLIBS_OPENCL ?= -framework OpenCL
+  else
+    LDLIBS_OPENCL ?= -lOpenCL
+  endif
+  OPENCL_DEVICE_ID ?= 0
+  OPENCL_PLATFORM_ID ?= 0
+
+  CPPFLAGS_OPENCL ?= -DSTAN_OPENCL -DOPENCL_DEVICE_ID=$(OPENCL_DEVICE_ID) -DOPENCL_PLATFORM_ID=$(OPENCL_PLATFORM_ID)
+  CXXFLAGS_OPENCL ?= -I $(OPENCL)
+endif
+
+################################################################################
+# Setup MPI
+#
+# Sets up MPI CXXFLAGS_MPI and LDLIBS_MPI to compile and link to MPI
+
+ifdef STAN_MPI
+  ifeq ($(OS),Windows_NT)
+    $(error MPI is not supported on Windows.)
+  endif
+
+  ifeq ($(OS),Darwin)
+    BOOST_LIBRARY_SUFFIX := .dylib
+  endif
+
+  ifeq ($(OS),Linux)
+    BOOST_LIBRARY_SUFFIX := .so
+  endif
+
+  MPI_TARGETS ?= $(addsuffix $(BOOST_LIBRARY_SUFFIX),$(BOOST)/stage/lib/libboost_serialization $(BOOST)/stage/lib/libboost_mpi) $(MPI_TEMPLATE_INSTANTIATION)
+  CPPFLAGS_MPI ?= -DSTAN_MPI
+
+  BOOST_LIBRARY_ABSOLUTE_PATH = $(abspath $(BOOST)/stage/lib)
+
+  LDFLAGS_MPI ?= -Wl,-L,"$(BOOST_LIBRARY_ABSOLUTE_PATH)" -Wl,-rpath,"$(BOOST_LIBRARY_ABSOLUTE_PATH)"
+  CXXFLAGS_MPI ?= -Wno-delete-non-virtual-dtor
```

Comment:
I curious where this no-delete-non-virtual-dtor comes from?

Reply:
It's a compiler warning that pops up. I think it was under Linux, but it might be under Mac. We should probably clear that up as a separate issue.